File: src/services/dynamodb/update.js
// services/dynamodb/update.js — DynamoDB UpdateExpression engine.
//
// Implements the four update clauses, in any order:
//   SET    path = operand [, ...]
//   REMOVE path [, ...]
//   ADD    path value [, ...]        (numbers add; sets union)
//   DELETE path value [, ...]        (sets difference)
//
// SET operands support arithmetic (a = a + :n, a = a - :n), the functions
// if_not_exists(path, operand) and list_append(op, op), placeholders :v, and
// document paths (nested maps + list indexes a.b[0].c). Names use #n.
//
// Items in MockCloud's store are unmarshalled JS. We therefore operate on JS
// values; raw ExpressionAttributeValues are unmarshalled to JS for the value
// side, while the original DynamoDB descriptor is consulted only to tell a Set
// (SS/NS/BS) apart from a List for ADD/DELETE semantics.

// ── Tokenizer (update grammar) ──────────────────────────────────────────────

function tokenize(src) {
  const tokens = [];
  let i = 0;
  while (i < src.length) {
    const c = src[i];
    if (c === ' ' || c === '\t' || c === '\n' || c === '\r') { i++; continue; }
    if (c === '(') { tokens.push({ t: 'lparen' }); i++; continue; }
    if (c === ')') { tokens.push({ t: 'rparen' }); i++; continue; }
    if (c === ',') { tokens.push({ t: 'comma' }); i++; continue; }
    if (c === '.') { tokens.push({ t: 'dot' }); i++; continue; }
    if (c === '=') { tokens.push({ t: 'eq' }); i++; continue; }
    if (c === '+') { tokens.push({ t: 'plus' }); i++; continue; }
    if (c === '-') { tokens.push({ t: 'minus' }); i++; continue; }
    if (c === '[') {
      let j = i + 1;
      while (j < src.length && src[j] !== ']') j++;
      if (j >= src.length) throw new Error('Unterminated list index');
      const n = src.slice(i + 1, j).trim();
      if (!/^\d+$/.test(n)) throw new Error(`Invalid list index: ${n}`);
      tokens.push({ t: 'index', v: Number(n) });
      i = j + 1; continue;
    }
    if (c === '#') {
      let j = i + 1;
      while (j < src.length && /[A-Za-z0-9_]/.test(src[j])) j++;
      tokens.push({ t: 'name_ph', v: src.slice(i, j) });
      i = j; continue;
    }
    if (c === ':') {
      let j = i + 1;
      while (j < src.length && /[A-Za-z0-9_]/.test(src[j])) j++;
      tokens.push({ t: 'value_ph', v: src.slice(i, j) });
      i = j; continue;
    }
    if (/[A-Za-z_]/.test(c)) {
      let j = i;
      while (j < src.length && /[A-Za-z0-9_]/.test(src[j])) j++;
      tokens.push({ t: 'ident', v: src.slice(i, j) });
      i = j; continue;
    }
    throw new Error(`Unexpected character in UpdateExpression: ${c}`);
  }
  return tokens;
}

// ── Clause splitting ────────────────────────────────────────────────────────

const CLAUSES = ['SET', 'REMOVE', 'ADD', 'DELETE'];

// Split the raw expression into { SET: '...', REMOVE: '...', ... } by locating
// the clause keywords as whole words (case-insensitive).
function splitClauses(expr) {
  const out = {};
  const re = /\b(SET|REMOVE|ADD|DELETE)\b/gi;
  const marks = [];
  let m;
  while ((m = re.exec(expr)) !== null) {
    marks.push({ kw: m[1].toUpperCase(), start: m.index, bodyStart: m.index + m[0].length });
  }
  for (let k = 0; k < marks.length; k++) {
    const end = k + 1 < marks.length ? marks[k + 1].start : expr.length;
    out[marks[k].kw] = expr.slice(marks[k].bodyStart, end).trim();
  }
  return out;
}

// ── Path helpers (operate on JS items) ──────────────────────────────────────

function resolveName(seg, names) {
  if (seg.kind === 'attr') return seg.v;
  if (seg.kind === 'name_ph') {
    if (!names || !(seg.v in names)) throw new Error(`Unknown name placeholder ${seg.v}`);
    return names[seg.v];
  }
  throw new Error('Index segment is not a name');
}

function getPath(item, segs, names) {
  let cur = item;
  for (const seg of segs) {
    if (cur === undefined || cur === null) return undefined;
    if (seg.kind === 'index') {
      if (!Array.isArray(cur)) return undefined;
      cur = cur[seg.v];
    } else {
      const name = resolveName(seg, names);
      if (typeof cur !== 'object' || Array.isArray(cur)) return undefined;
      cur = cur[name];
    }
  }
  return cur;
}

function setPath(item, segs, value, names) {
  let cur = item;
  for (let k = 0; k < segs.length; k++) {
    const seg = segs[k];
    const last = k === segs.length - 1;
    const key = seg.kind === 'index' ? seg.v : resolveName(seg, names);
    if (last) { cur[key] = value; return; }
    const nextIsIndex = segs[k + 1].kind === 'index';
    if (cur[key] === undefined || cur[key] === null) cur[key] = nextIsIndex ? [] : {};
    cur = cur[key];
  }
}

function removePath(item, segs, names) {
  let cur = item;
  for (let k = 0; k < segs.length; k++) {
    const seg = segs[k];
    const last = k === segs.length - 1;
    const key = seg.kind === 'index' ? seg.v : resolveName(seg, names);
    if (cur === undefined || cur === null) return;
    if (last) {
      if (seg.kind === 'index') { if (Array.isArray(cur)) cur.splice(key, 1); }
      else if (typeof cur === 'object') delete cur[key];
      return;
    }
    cur = cur[key];
  }
}

// ── Token-stream parser ─────────────────────────────────────────────────────
//
// We parse each clause body into actions. Reusable cursor over a token array.

function makeCursor(tokens) {
  let pos = 0;
  return {
    peek: (o = 0) => tokens[pos + o],
    next: () => tokens[pos++],
    eof: () => pos >= tokens.length,
    expect(t) {
      const tok = tokens[pos];
      if (!tok || tok.t !== t) throw new Error(`Expected ${t}, got ${tok ? tok.t : 'EOF'}`);
      return tokens[pos++];
    },
  };
}

function parsePath(cur) {
  const segs = [];
  const head = cur.next();
  if (!head) throw new Error('Expected path');
  if (head.t === 'name_ph') segs.push({ kind: 'name_ph', v: head.v });
  else if (head.t === 'ident') segs.push({ kind: 'attr', v: head.v });
  else throw new Error(`Expected path, got ${head.t}`);
  while (!cur.eof()) {
    const nx = cur.peek();
    if (nx.t === 'dot') {
      cur.next();
      const part = cur.next();
      if (part.t === 'name_ph') segs.push({ kind: 'name_ph', v: part.v });
      else if (part.t === 'ident') segs.push({ kind: 'attr', v: part.v });
      else throw new Error('Expected name after "."');
    } else if (nx.t === 'index') {
      cur.next();
      segs.push({ kind: 'index', v: nx.v });
    } else break;
  }
  return segs;
}

// ── SET operand evaluation ──────────────────────────────────────────────────
//
// operand := term (('+' | '-') term)?
// term    := value_ph | path | if_not_exists(path, operand) | list_append(op, op)

function evalOperand(cur, item, names, vals) {
  let left = evalTerm(cur, item, names, vals);
  const nx = cur.peek();
  if (nx && (nx.t === 'plus' || nx.t === 'minus')) {
    cur.next();
    const right = evalTerm(cur, item, names, vals);
    const a = Number(left), b = Number(right);
    if (Number.isNaN(a) || Number.isNaN(b)) {
      throw new Error('Arithmetic on non-numeric operand in UpdateExpression');
    }
    return nx.t === 'plus' ? a + b : a - b;
  }
  return left;
}

function evalTerm(cur, item, names, vals) {
  const tok = cur.peek();
  if (!tok) throw new Error('Expected operand in SET');
  if (tok.t === 'value_ph') {
    cur.next();
    if (!(tok.v in vals)) throw new Error(`Unknown value placeholder ${tok.v}`);
    return vals[tok.v].js;
  }
  if (tok.t === 'ident' && cur.peek(1) && cur.peek(1).t === 'lparen') {
    const fn = tok.v.toLowerCase();
    if (fn === 'if_not_exists') {
      cur.next(); cur.next(); // ident, lparen
      const segs = parsePath(cur);
      cur.expect('comma');
      const fallback = evalOperand(cur, item, names, vals);
      cur.expect('rparen');
      const existing = getPath(item, segs, names);
      return existing === undefined ? fallback : existing;
    }
    if (fn === 'list_append') {
      cur.next(); cur.next();
      const a = evalOperand(cur, item, names, vals);
      cur.expect('comma');
      const b = evalOperand(cur, item, names, vals);
      cur.expect('rparen');
      const la = Array.isArray(a) ? a : (a === undefined ? [] : [a]);
      const lb = Array.isArray(b) ? b : (b === undefined ? [] : [b]);
      return [...la, ...lb];
    }
    throw new Error(`Unknown function in SET: ${fn}`);
  }
  // path
  const segs = parsePath(cur);
  return getPath(item, segs, names);
}

// ── Set helpers for ADD / DELETE ────────────────────────────────────────────

function asArray(v) { return Array.isArray(v) ? v.slice() : []; }
function unionSet(a, b) {
  const out = asArray(a);
  for (const x of b) if (!out.some(y => eq(y, x))) out.push(x);
  return out;
}
function diffSet(a, b) { return asArray(a).filter(x => !b.some(y => eq(y, x))); }
function eq(a, b) { return a === b || JSON.stringify(a) === JSON.stringify(b); }

// ── Public entry ────────────────────────────────────────────────────────────
//
// applyUpdate(oldItem, updateExpr, exprNames, rawValues, jsValues)
//   oldItem   — current stored item (JS) or null
//   updateExpr— the UpdateExpression string
//   exprNames — ExpressionAttributeNames map (#n → real name)
//   rawValues — raw ExpressionAttributeValues (type-descriptor shaped) — used
//               to distinguish sets from lists for ADD/DELETE
//   jsValues  — the same values unmarshalled to JS
//
// Returns { item, changed } where `item` is the new image (a fresh object) and
// `changed` is the array of top-level attribute names touched (for ReturnValues
// UPDATED_OLD / UPDATED_NEW).
export function applyUpdate(oldItem, updateExpr, exprNames, rawValues, jsValues) {
  const names = exprNames || {};
  const vals = {};
  for (const [k, v] of Object.entries(jsValues || {})) {
    const raw = rawValues ? rawValues[k] : undefined;
    const isSet = !!(raw && ('SS' in raw || 'NS' in raw || 'BS' in raw));
    // Storage is unmarshalled JS and the generic unmarshal() does not handle
    // set descriptors — extract the members array here so ADD/DELETE see an
    // array. (Sets degrade to lists in the lossy Phase-1 storage model.)
    let js = v;
    if (isSet) {
      if ('SS' in raw) js = raw.SS.slice();
      else if ('NS' in raw) js = raw.NS.map(Number);
      else js = raw.BS.slice();
    }
    vals[k] = { js, raw, isSet };
  }

  const item = oldItem ? structuredCloneSafe(oldItem) : {};
  const changed = new Set();
  const clauses = splitClauses(updateExpr || '');

  if (clauses.SET) {
    const cur = makeCursor(tokenize(clauses.SET));
    while (!cur.eof()) {
      const segs = parsePath(cur);
      cur.expect('eq');
      const value = evalOperand(cur, item, names, vals);
      setPath(item, segs, value, names);
      changed.add(topName(segs, names));
      if (!cur.eof()) cur.expect('comma');
    }
  }

  if (clauses.REMOVE) {
    const cur = makeCursor(tokenize(clauses.REMOVE));
    while (!cur.eof()) {
      const segs = parsePath(cur);
      removePath(item, segs, names);
      changed.add(topName(segs, names));
      if (!cur.eof()) cur.expect('comma');
    }
  }

  if (clauses.ADD) {
    const cur = makeCursor(tokenize(clauses.ADD));
    while (!cur.eof()) {
      const segs = parsePath(cur);
      const tok = cur.expect('value_ph');
      if (!(tok.v in vals)) throw new Error(`Unknown value placeholder ${tok.v}`);
      const operand = vals[tok.v];
      const existing = getPath(item, segs, names);
      if (operand.isSet) {
        setPath(item, segs, unionSet(existing, operand.js), names);
      } else {
        const base = existing === undefined ? 0 : Number(existing);
        setPath(item, segs, base + Number(operand.js), names);
      }
      changed.add(topName(segs, names));
      if (!cur.eof()) cur.expect('comma');
    }
  }

  if (clauses.DELETE) {
    const cur = makeCursor(tokenize(clauses.DELETE));
    while (!cur.eof()) {
      const segs = parsePath(cur);
      const tok = cur.expect('value_ph');
      if (!(tok.v in vals)) throw new Error(`Unknown value placeholder ${tok.v}`);
      const operand = vals[tok.v];
      const existing = getPath(item, segs, names);
      const result = diffSet(existing, operand.js);
      if (result.length === 0) removePath(item, segs, names);
      else setPath(item, segs, result, names);
      changed.add(topName(segs, names));
      if (!cur.eof()) cur.expect('comma');
    }
  }

  return { item, changed: [...changed] };
}

function topName(segs, names) {
  return segs[0].kind === 'index' ? null : resolveName(segs[0], names);
}

function structuredCloneSafe(obj) {
  return JSON.parse(JSON.stringify(obj));
}

File: internal/services/dynamodb/update.go
// update.go — 1:1 port of src/services/dynamodb/update.js: the
// UpdateExpression engine (SET / REMOVE / ADD / DELETE).
//
// Fidelity notes:
//   - clauses split by a whole-word case-insensitive regex on the RAW string
//     (duplicate SET → last occurrence wins);
//   - if_not_exists falls back ONLY when the path is missing (stored JSON
//     null counts as existing);
//   - SET arithmetic uses JS Number coercion (Number(null)===0, so
//     arithmetic over JSON-null succeeds; missing → NaN → throws);
//   - ADD decides set-vs-number from the RAW value descriptor (SS|NS|BS);
//   - DELETE with an empty result removes the attribute;
//   - list_append wraps non-arrays;
//   - `changed` = top-level names in clause order (index-headed paths → null,
//     which pickAttrs skips).
package dynamodb

import (
	"math"
	"regexp"
	"strings"

	"github.com/mockcloud/mockcloud/internal/jsnum"
)

// ── Tokenizer (update grammar) ──────────────────────────────────────────────

func updTokenize(src string) []tok {
	var tokens []tok
	rs := []rune(src)
	i := 0
	for i < len(rs) {
		c := rs[i]
		switch {
		case c == ' ' || c == '\t' || c == '\n' || c == '\r':
			i++
		case c == '(':
			tokens = append(tokens, mkTok("lparen"))
			i++
		case c == ')':
			tokens = append(tokens, mkTok("rparen"))
			i++
		case c == ',':
			tokens = append(tokens, mkTok("comma"))
			i++
		case c == '.':
			tokens = append(tokens, mkTok("dot"))
			i++
		case c == '=':
			tokens = append(tokens, mkTok("eq"))
			i++
		case c == '+':
			tokens = append(tokens, mkTok("plus"))
			i++
		case c == '-':
			tokens = append(tokens, mkTok("minus"))
			i++
		case c == '[':
			j := i + 1
			for j < len(rs) && rs[j] != ']' {
				j++
			}
			if j >= len(rs) {
				throwf("Unterminated list index")
			}
			n := strings.TrimSpace(string(rs[i+1 : j]))
			if !isDigits(n) {
				throwf("Invalid list index: %s", n)
			}
			tokens = append(tokens, mkTokV("index", jsnum.ToNumberFromString(n)))
			i = j + 1
		case c == '#':
			j := i + 1
			for j < len(rs) && isWordChar(rs[j]) {
				j++
			}
			tokens = append(tokens, mkTokV("name_ph", string(rs[i:j])))
			i = j
		case c == ':':
			j := i + 1
			for j < len(rs) && isWordChar(rs[j]) {
				j++
			}
			tokens = append(tokens, mkTokV("value_ph", string(rs[i:j])))
			i = j
		case isWordStart(c):
			j := i
			for j < len(rs) && isWordChar(rs[j]) {
				j++
			}
			tokens = append(tokens, mkTokV("ident", string(rs[i:j])))
			i = j
		default:
			throwf("Unexpected character in UpdateExpression: %c", c)
		}
	}
	return tokens
}

// ── Clause splitting ────────────────────────────────────────────────────────

var clauseRe = regexp.MustCompile(`(?i)\b(SET|REMOVE|ADD|DELETE)\b`)

// splitClauses locates the clause keywords as whole words (case-insensitive)
// in the raw string. Later duplicates overwrite earlier ones.
func splitClauses(expr string) map[string]string {
	out := map[string]string{}
	locs := clauseRe.FindAllStringSubmatchIndex(expr, -1)
	for k, m := range locs {
		kw := strings.ToUpper(expr[m[2]:m[3]])
		bodyStart := m[1]
		end := len(expr)
		if k+1 < len(locs) {
			end = locs[k+1][0]
		}
		out[kw] = strings.TrimSpace(expr[bodyStart:end])
	}
	return out
}

// ── Path helpers (operate on JS items) ──────────────────────────────────────

// getPath — update.js getPath (same walk as the expression engine's).
func getPath(item map[string]any, segs []seg, names map[string]any) any {
	return walkPath(segs, item, names)
}

// setPath — update.js setPath (delegates to the shared materialiser).
func setPath(item map[string]any, segs []seg, value any, names map[string]any) {
	setInto(item, segs, value, names)
}

// removePath — update.js removePath.
func removePath(item map[string]any, segs []seg, names map[string]any) {
	var cur any = item
	for k := 0; k < len(segs); k++ {
		sg := segs[k]
		last := k == len(segs)-1
		var key string
		if sg.kind != "index" {
			key = resolveName(sg, names)
		}
		if cur == nil || jsnum.IsUndef(cur) {
			return
		}
		if last {
			if sg.kind == "index" {
				if arr, ok := cur.([]any); ok && sg.idx >= 0 && sg.idx < len(arr) {
					// splice(idx, 1) mutates in place — copy down and shrink,
					// then propagate: arrays are mutated via the parent
					// reference in JS; we mirror by rewriting through parents.
					copy(arr[sg.idx:], arr[sg.idx+1:])
					shrinkParent(item, segs[:k], arr[:len(arr)-1], names)
				}
			} else if m, ok := cur.(map[string]any); ok {
				delete(m, key)
			}
			return
		}
		switch c := cur.(type) {
		case map[string]any:
			if sg.kind == "index" {
				cur = attrVal(c, jsnum.Format(float64(sg.idx)))
			} else {
				cur = attrVal(c, key)
			}
		case []any:
			if sg.kind == "index" && sg.idx >= 0 && sg.idx < len(c) {
				cur = c[sg.idx]
			} else {
				cur = jsnum.Undef
			}
		default:
			cur = jsnum.Undef
		}
	}
}

// shrinkParent re-assigns a shortened slice into its parent container (Go
// slices can't shrink in place through an interface value).
func shrinkParent(item map[string]any, parentSegs []seg, shrunk []any, names map[string]any) {
	if len(parentSegs) == 0 {
		return // the item root is a map; an index-headed path can't reach here
	}
	setInto(item, parentSegs, shrunk, names)
}

// ── Token-stream cursor ─────────────────────────────────────────────────────

type cursor struct {
	tokens []tok
	pos    int
}

func (c *cursor) peek(off int) *tok {
	if c.pos+off < len(c.tokens) {
		return &c.tokens[c.pos+off]
	}
	return nil
}
func (c *cursor) next() *tok {
	t := c.peek(0)
	c.pos++
	return t
}
func (c *cursor) eof() bool { return c.pos >= len(c.tokens) }
func (c *cursor) expect(t string) *tok {
	tk := c.peek(0)
	if tk == nil || tk.t != t {
		got := "EOF"
		if tk != nil {
			got = tk.t
		}
		throwf("Expected %s, got %s", t, got)
	}
	c.pos++
	return tk
}

func updParsePath(c *cursor) []seg {
	var segs []seg
	head := c.next()
	if head == nil {
		throwf("Expected path")
	}
	switch head.t {
	case "name_ph":
		segs = append(segs, seg{kind: "name_ph", name: head.v.(string)})
	case "ident":
		segs = append(segs, seg{kind: "attr", name: head.v.(string)})
	default:
		throwf("Expected path, got %s", head.t)
	}
	for !c.eof() {
		nx := c.peek(0)
		if nx.t == "dot" {
			c.next()
			part := c.next()
			if part == nil {
				throwf(`Expected name after "."`)
			} else if part.t == "name_ph" {
				segs = append(segs, seg{kind: "name_ph", name: part.v.(string)})
			} else if part.t == "ident" {
				segs = append(segs, seg{kind: "attr", name: part.v.(string)})
			} else {
				throwf(`Expected name after "."`)
			}
		} else if nx.t == "index" {
			c.next()
			segs = append(segs, seg{kind: "index", idx: int(nx.v.(float64))})
		} else {
			break
		}
	}
	return segs
}

// ── SET operand evaluation ──────────────────────────────────────────────────

type updVal struct {
	js    any
	raw   map[string]any
	isSet bool
}

func evalUpdOperand(c *cursor, item map[string]any, names map[string]any, vals map[string]updVal) any {
	left := evalUpdTerm(c, item, names, vals)
	nx := c.peek(0)
	if nx != nil && (nx.t == "plus" || nx.t == "minus") {
		c.next()
		right := evalUpdTerm(c, item, names, vals)
		a, b := jsnum.ToNumber(left), jsnum.ToNumber(right)
		if math.IsNaN(a) || math.IsNaN(b) {
			throwf("Arithmetic on non-numeric operand in UpdateExpression")
		}
		if nx.t == "plus" {
			return a + b
		}
		return a - b
	}
	return left
}

func evalUpdTerm(c *cursor, item map[string]any, names map[string]any, vals map[string]updVal) any {
	tk := c.peek(0)
	if tk == nil {
		throwf("Expected operand in SET")
	}
	if tk.t == "value_ph" {
		c.next()
		v, ok := vals[tk.v.(string)]
		if !ok {
			throwf("Unknown value placeholder %s", tk.v.(string))
		}
		return v.js
	}
	if tk.t == "ident" {
		if n1 := c.peek(1); n1 != nil && n1.t == "lparen" {
			fn := strings.ToLower(tk.v.(string))
			switch fn {
			case "if_not_exists":
				c.next()
				c.next() // ident, lparen
				segs := updParsePath(c)
				c.expect("comma")
				fallback := evalUpdOperand(c, item, names, vals)
				c.expect("rparen")
				existing := getPath(item, segs, names)
				if jsnum.IsUndef(existing) {
					return fallback
				}
				return existing
			case "list_append":
				c.next()
				c.next()
				a := evalUpdOperand(c, item, names, vals)
				c.expect("comma")
				b := evalUpdOperand(c, item, names, vals)
				c.expect("rparen")
				la := wrapList(a)
				lb := wrapList(b)
				out := make([]any, 0, len(la)+len(lb))
				out = append(out, la...)
				out = append(out, lb...)
				return out
			default:
				throwf("Unknown function in SET: %s", fn)
			}
		}
	}
	segs := updParsePath(c)
	return getPath(item, segs, names)
}

// wrapList — Array.isArray(a) ? a : (a === undefined ? [] : [a]).
func wrapList(v any) []any {
	if arr, ok := v.([]any); ok {
		return arr
	}
	if jsnum.IsUndef(v) {
		return []any{}
	}
	return []any{v}
}

// ── Set helpers for ADD / DELETE ────────────────────────────────────────────

func asArray(v any) []any {
	if arr, ok := v.([]any); ok {
		out := make([]any, len(arr))
		copy(out, arr)
		return out
	}
	return []any{}
}

// updEq — a === b || JSON.stringify(a) === JSON.stringify(b).
func updEq(a, b any) bool {
	if jsnum.StrictEq(a, b) {
		return true
	}
	sa, aok := stringifyJSON(a)
	sb, bok := stringifyJSON(b)
	return aok && bok && sa == sb
}

func unionSet(a any, b []any) []any {
	out := asArray(a)
	for _, x := range b {
		found := false
		for _, y := range out {
			if updEq(y, x) {
				found = true
				break
			}
		}
		if !found {
			out = append(out, x)
		}
	}
	return out
}

// diffSet — asArray(a).filter(x => !b.some(y => eq(y, x))). The `b.some`
// crash on a non-array operand only fires when `a` has elements, exactly like
// the JS filter callback.
func diffSet(a any, b any) []any {
	out := []any{}
	for _, x := range asArray(a) {
		barr, ok := b.([]any)
		if !ok {
			throwf("b.some is not a function")
		}
		found := false
		for _, y := range barr {
			if updEq(y, x) {
				found = true
				break
			}
		}
		if !found {
			out = append(out, x)
		}
	}
	return out
}

// ── Public entry ────────────────────────────────────────────────────────────

type updateResult struct {
	item    map[string]any
	changed []string
}

// applyUpdate — see update.js applyUpdate. rawValues distinguishes sets from
// lists; jsValues is the same map unmarshalled to JS.
func applyUpdate(oldItem map[string]any, updateExpr string, exprNames map[string]any,
	rawValues map[string]any, jsValues map[string]any) (res updateResult, err error) {
	defer catchJS(&err)

	names := exprNames
	vals := map[string]updVal{}
	for k, v := range jsValues {
		var raw map[string]any
		if rawValues != nil {
			raw, _ = rawValues[k].(map[string]any)
		}
		isSet := false
		if raw != nil {
			_, ss := raw["SS"]
			_, ns := raw["NS"]
			_, bs := raw["BS"]
			isSet = ss || ns || bs
		}
		js := v
		if isSet {
			if ssv, has := raw["SS"]; has {
				js = asArray(ssv)
			} else if nsv, has := raw["NS"]; has {
				arr := asArray(nsv)
				out := make([]any, len(arr))
				for i, e := range arr {
					out[i] = jsnum.ToNumber(e)
				}
				js = out
			} else {
				js = asArray(raw["BS"])
			}
		}
		vals[k] = updVal{js: js, raw: raw, isSet: isSet}
	}

	var item map[string]any
	if oldItem != nil {
		item = jsonCloneMap(oldItem)
	} else {
		item = map[string]any{}
	}
	var changed []string
	seen := map[string]bool{}
	addChanged := func(name string) {
		if !seen[name] {
			seen[name] = true
			changed = append(changed, name)
		}
	}
	clauses := splitClauses(updateExpr)

	if body := clauses["SET"]; body != "" {
		c := &cursor{tokens: updTokenize(body)}
		for !c.eof() {
			segs := updParsePath(c)
			c.expect("eq")
			value := evalUpdOperand(c, item, names, vals)
			setPath(item, segs, value, names)
			addChanged(topName(segs, names))
			if !c.eof() {
				c.expect("comma")
			}
		}
	}

	if body := clauses["REMOVE"]; body != "" {
		c := &cursor{tokens: updTokenize(body)}
		for !c.eof() {
			segs := updParsePath(c)
			removePath(item, segs, names)
			addChanged(topName(segs, names))
			if !c.eof() {
				c.expect("comma")
			}
		}
	}

	if body := clauses["ADD"]; body != "" {
		c := &cursor{tokens: updTokenize(body)}
		for !c.eof() {
			segs := updParsePath(c)
			tk := c.expect("value_ph")
			operand, ok := vals[tk.v.(string)]
			if !ok {
				throwf("Unknown value placeholder %s", tk.v.(string))
			}
			existing := getPath(item, segs, names)
			if operand.isSet {
				members, _ := operand.js.([]any) // isSet ⇒ built as []any above
				setPath(item, segs, unionSet(existing, members), names)
			} else {
				base := 0.0
				if !jsnum.IsUndef(existing) {
					base = jsnum.ToNumber(existing)
				}
				setPath(item, segs, base+jsnum.ToNumber(operand.js), names)
			}
			addChanged(topName(segs, names))
			if !c.eof() {
				c.expect("comma")
			}
		}
	}

	if body := clauses["DELETE"]; body != "" {
		c := &cursor{tokens: updTokenize(body)}
		for !c.eof() {
			segs := updParsePath(c)
			tk := c.expect("value_ph")
			operand, ok := vals[tk.v.(string)]
			if !ok {
				throwf("Unknown value placeholder %s", tk.v.(string))
			}
			existing := getPath(item, segs, names)
			result := diffSet(existing, operand.js)
			if len(result) == 0 {
				removePath(item, segs, names)
			} else {
				setPath(item, segs, result, names)
			}
			addChanged(topName(segs, names))
			if !c.eof() {
				c.expect("comma")
			}
		}
	}

	return updateResult{item: item, changed: changed}, nil
}

// topName — segs[0] index-headed → "" (JS null; both are skipped by
// pickAttrs' falsy check).
func topName(segs []seg, names map[string]any) string {
	if segs[0].kind == "index" {
		return ""
	}
	return resolveName(segs[0], names)
}
